r1cs/gadgets/std/src/fields/nonnative/tests.rs

```diff
@@ -1,365 +1,1120 @@
+                (ConstraintF::Params::CAPACITY as usize - 3 - params.bits_per_limb) as u32
+            ) - BigUint::from(2u32)
+        ).bits() as usize - 1;
+        
+        let surfeit_a = rng.gen_range(0..=surfeit_bound);
+        let surfeit_b = rng.gen_range(0..=surfeit_bound);
+        
+        let a = NonNativeFieldGadget::<SimulationF, ConstraintF>::alloc_random(
+            cs.ns(|| "alloc random a" ),
+            rng, 
+            surfeit_a).unwrap();
+
+        assert!(
+            a.check(), 
+            "random allocated a fails on check()"
+        );
+
+        let a_native = a.get_value().unwrap();
+
+        let b = NonNativeFieldGadget::<SimulationF, ConstraintF>::alloc_random(
+            cs.ns(|| "alloc random b" ),
+            rng, 
+            surfeit_b).unwrap();
+
+        assert!(
+            b.check(), 
+            "random allocated b fails on check()"
+        );
+
+        let b_native = b.get_value().unwrap();
+
+        let a_times_b = a.mul(cs.ns(|| "a * b"), &b).unwrap();
+        let b_times_a = b.mul(cs.ns(|| "b * a"), &a).unwrap();
+
+        let a_times_b_actual = a_times_b.get_value().unwrap();
+        let a_times_b_expected = a_native * &b_native;
+        let b_times_a_actual = b_times_a.get_value().unwrap();
+
+        assert!(
+            a_times_b_actual.eq(&a_times_b_expected),
+            "a_times_b = {:?}, a_times_b_actual = {:?}, a_times_b_expected = {:?}",
+            a_times_b,
+            a_times_b_actual.into_repr().as_ref(),
+            a_times_b_expected.into_repr().as_ref()
+        );
+
+        assert!(
+            b_times_a_actual.eq(&a_times_b_expected),
+            "b_times_a = {:?}, b_times_a_actual = {:?}, a_times_b_expected = {:?}",
+            b_times_a,
+            b_times_a_actual.into_repr().as_ref(),
+            a_times_b_expected.into_repr().as_ref()
+        );
+
+        if !cs.is_satisfied() {
+            println!("{:?}", cs.which_is_unsatisfied());
+        }
+        assert!(cs.is_satisfied());
+    }
+}
+
+fn elementary_test_multiplication_by_constant<SimulationF: PrimeField, ConstraintF: PrimeField, R: RngCore>(rng: &mut R) {
+    for i in 0..TEST_COUNT {
+        let mut cs = ConstraintSystem::<ConstraintF>::new(SynthesisMode::Debug);
+        let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
+
+        // We sample reducible nonnatives.        
+        // ``
+        //      2^surfeit + 2 <= 2^{CAPACITY - 3 - bits_per_limb}.
+        // ``
+        // Notice that `floor_log_2(x) = len(x) - 1`.
+        // Edge cases of mul_without_prereduce() are characterized by
+        // ``
+        //    num_limbs^2 * (num_add + 1)  + 1  
+        //                          <= 2^{CAPACITY - 2 - 2*bits_per_limb},
+        // ``
+        // or
+        // ``
+        //    num_limbs^2 * 2^surfeit_a  + 1  
+        //                          <= 2^{CAPACITY - 2 - 2*bits_per_limb},
+        // ``
```

Comment:
Agree!

Reply:
The part of the comment related to `mul_without_prereduce` will be removed in PR #164, as those bounds are not employed in this test (`mul_by_constant` is employed rather than `mul_without_prereduce`)